impala/hiveserver2.py

```diff
@@ -1112,37 +1143,43 @@ def __init__(self, service, handle, config, hs2_protocol_version,
 
     def close(self):
         req = TCloseSessionReq(sessionHandle=self.handle)
-        self._rpc('CloseSession', req)
+        # CloseSession rpcs don't need retries since we catch all exceptions and close
+        # transport.
+        self._rpc('CloseSession', req, False)
 
     def execute(self, statement, configuration=None, run_async=False):
         req = TExecuteStatementReq(sessionHandle=self.handle,
                                    statement=statement,
                                    confOverlay=configuration,
                                    runAsync=run_async)
-        return self._operation('ExecuteStatement', req)
+        # Do not try to retry http requests.
+        # Read queries should be idempotent but most dml queries are not. Also retrying
+        # query execution from client could be expensive and so likely makes sense to do
+        # it if server is also aware of the retries.
+        return self._operation('ExecuteStatement', req, False)
 
     def get_databases(self, schema='.*'):
         req = TGetSchemasReq(sessionHandle=self.handle, schemaName=schema)
-        return self._operation('GetSchemas', req)
+        return self._operation('GetSchemas', req, False)
```

Comment:
Typically all the read operations such as GetSchemas, GetTables, GetColumns, GetFunctions should be idempotent, as long as they return all the results in a single call - it's either all or nothing and so retrying these functions should be okay.

But, I am okay if you prefer to limit the scope of retries to the more common functions at this time.

Reply:
I'll change the code for these 4 functions